contrib/errorprone/src/python/pants/contrib/errorprone/tasks/errorprone.py

```diff
@@ -0,0 +1,140 @@
+                          ],
+                          main=cls._ERRORPRONE_MAIN,
+                          custom_rules=[
+                            Shader.exclude_package('com.google.errorprone', recursive=True),
+                          ]
+                          )
+
+  @classmethod
+  def prepare(cls, options, round_manager):
+    super(ErrorProne, cls).prepare(options, round_manager)
+    round_manager.require_data('runtime_classpath')
+
+  def _is_errorprone_target(self, target):
+    return target.has_sources(self._JAVA_SOURCE_EXTENSION) and not target.is_synthetic
+
+  @property
+  def cache_target_dirs(self):
+    return True
+
+  def execute(self):
+    if self.get_options().skip:
+      return
+
+    invalidate_deps = False
+    if self.get_options().transitive:
+      targets = self.context.targets(self._is_errorprone_target)
+      invalidate_deps = True
+    else:
+      targets = filter(self._is_errorprone_target, self.context.target_roots)
+
+    targets = list(set(targets))
+
+    target_count = 0
+    errorprone_failed = False
+    with self.invalidated(targets, invalidate_dependents=invalidate_deps) as invalidation_check:
+      total_targets = len(invalidation_check.invalid_vts)
+      for vt in invalidation_check.invalid_vts:
+        target_count += 1
+        self.context.log.info('[{}/{}] {}'.format(str(target_count).rjust(len(str(total_targets))), total_targets, vt.target.address.spec))
+
+        if vt.target.address.spec in self.get_options().excluded_targets:
+          self.context.log.info('  Skipping [{}] because it is an excluded target'.format(vt.target.address.spec))
+          vt.update()
+          continue
+
+        sources = self.calculate_sources(vt.target)
+        if sources:
+          result = self.errorprone(vt.target, sources)
+          if result != 0:
+            if self.get_options().fail_fast:
+              raise TaskError('java {main} ... exited non-zero ({result})'.format(
+                main=self._ERRORPRONE_MAIN, result=result))
+            else:
+              errorprone_failed = True
+          else:
+            vt.update()
+        else:
+          self.context.log.info('  Skipping [{}] because it contains no java files'.format(vt.target.address.spec))
+          vt.update()
+
+      if errorprone_failed:
+        raise TaskError('ErrorProne checks failed')
+
+  def calculate_sources(self, target):
+    sources = set()
+    sources.update(source for source in target.sources_relative_to_buildroot()
+                   if source.endswith(self._JAVA_SOURCE_EXTENSION))
+    return sources
+
+  def errorprone(self, target, sources):
+    union_classpath = OrderedSet(self.tool_classpath('errorprone'))
+    runtime_classpaths = self.context.products.get_data('runtime_classpath')
+    runtime_classpath = runtime_classpaths.get_for_targets(target.closure(bfs=True))
+    union_classpath.update(jar for conf, jar in runtime_classpath)
```

Comment:
Looks like we add the checkstyle jars to the classpath for checkstyle as well.

In Ant, Maven and Gradle there is plugin infrastructure that lets you specify a separate classpath for the plugin dependencies which are often managed by using a separate classloader for the plugin.  We don't really have something like that in Pants.  We could look at adding something to nailgun_task.py or java/util.py to allow it though?  I think it would mean creating our own java class that took arguments for the plugin classpath and tool main class and created a plugin classloader separate from the runtime classpath.